File: tests/component_integration/endpoints/test_audio_transcription_endpoint.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Component integration tests for the audio_transcription endpoint type."""

import pytest

from aiperf.common.constants import IS_WINDOWS_ARM
from tests.component_integration.conftest import (
    ComponentIntegrationTestDefaults as defaults,
)
from tests.harness.utils import AIPerfCLI


@pytest.mark.component_integration
@pytest.mark.skipif(
    IS_WINDOWS_ARM,
    reason="synthetic audio generation requires libsndfile, which has no "
    "prebuilt Windows-on-ARM wheel (see audio.py::import_soundfile)",
)
class TestAudioTranscriptionEndpoint:
    """Smoke tests for the audio_transcription endpoint against the mock server."""

    def test_synthetic_audio_transcription(self, cli: AIPerfCLI) -> None:
        """Audio transcription with synthetically generated WAV audio."""
        result = cli.run_sync(
            f"""
            aiperf profile \
                --model openai/whisper-large-v3 \
                --endpoint-type audio_transcription \
                --audio-batch-size 1 \
                --audio-length-mean 3.0 \
                --audio-length-stddev 0.0 \
                --request-count {defaults.request_count} \
                --concurrency {defaults.concurrency} \
                --workers-max {defaults.workers_max} \
                --ui {defaults.ui}
            """
        )
        assert result.request_count == defaults.request_count

    def test_audio_transcription_with_extra_inputs(self, cli: AIPerfCLI) -> None:
        """--extra-inputs (language, temperature) flow through to multipart form fields."""
        result = cli.run_sync(
            f"""
            aiperf profile \
                --model openai/whisper-large-v3 \
                --endpoint-type audio_transcription \
                --audio-batch-size 1 \
                --audio-length-mean 3.0 \
                --audio-length-stddev 0.0 \
                --extra-inputs language:en temperature:0.0 \
                --request-count {defaults.request_count} \
                --concurrency {defaults.concurrency} \
                --workers-max {defaults.workers_max} \
                --ui {defaults.ui}
            """
        )
        assert result.request_count == defaults.request_count

File: rust/e2e-tests/tests/test_audio_transcription_endpoint.rs
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

mod common;
use common::*;

use serde_json::Value;

fn non_streaming_body(record: &Value) -> Option<Value> {
    let responses = record.get("responses").and_then(Value::as_array)?;
    for resp in responses {
        if let Some(text) = resp.get("text").and_then(Value::as_str) {
            if let Ok(body) = serde_json::from_str::<Value>(text) {
                return Some(body);
            }
        }
    }
    None
}

#[tokio::test]
async fn test_audio_transcription_passes_generated_audio_and_extra_inputs() {
    let h = AIPerfHarness::new().await;
    let r = h.run(&format!(
        "--model openai/whisper-1 \
         --tokenizer gpt2 \
         --url {} \
         --endpoint-type audio_transcription \
         --audio-length-mean 0.1 \
         --audio-format wav \
         --extra-inputs language:en temperature:0.0 \
         --request-count 4 \
         --concurrency 2 \
         --workers-max 1 \
         --export-level raw \
         --ui simple",
        h.mock.url
    ));
    assert!(
        r.success(),
        "audio transcription run failed (exit {}):\nstdout:\n{}\nstderr:\n{}",
        r.exit_code,
        r.stdout,
        r.stderr
    );
    assert_eq!(r.artifacts.request_count() as u32, 4);

    let records = r.artifacts.raw_records();
    assert_eq!(records.len(), 4, "expected 4 raw records");
    for (index, record) in records.iter().enumerate() {
        assert_eq!(
            record.get("status").and_then(Value::as_u64),
            Some(200),
            "record {index}: status not 200\n{record}"
        );
        assert!(
            record["request_headers"]["Content-Type"]
                .as_str()
                .is_some_and(|value| value.starts_with("multipart/form-data; boundary=")),
            "record {index}: request did not go over native multipart\n{record}"
        );
        let body = non_streaming_body(record)
            .unwrap_or_else(|| panic!("record {index}: no parseable response body\n{record}"));
        assert_eq!(body["text"], "mock transcription", "record {index}");
        assert_eq!(body["language"], "en", "record {index}");
        assert_eq!(body["temperature"], 0.0, "record {index}");
        assert_eq!(body["usage"]["input_tokens"], 1, "record {index}");
    }
}
